volume/volume.go

```diff
@@ -25,33 +35,93 @@ type Volume interface {
 	Unmount() error
 }
 
-// read-write modes
-var rwModes = map[string]bool{
-	"rw":   true,
-	"rw,Z": true,
-	"rw,z": true,
-	"z,rw": true,
-	"Z,rw": true,
-	"Z":    true,
-	"z":    true,
+var validModeFlags = map[string]bool{
+	"ro": true,
+	"rw": true,
+	"u":  true,
+	"z":  true,
+	"Z":  true,
+}
+
+func validMode(modes string) bool {
+	arr := strings.Split(modes, ",")
+	for _, mode := range arr {
+		if !validModeFlags[mode] {
+			return false
+		}
+	}
+	return true
 }
 
-// read-only modes
-var roModes = map[string]bool{
-	"ro":   true,
-	"ro,Z": true,
-	"ro,z": true,
-	"z,ro": true,
-	"Z,ro": true,
+func rwMode(modes string) bool {
+	arr := strings.Split(modes, ",")
+	for _, mode := range arr {
+		if mode == "ro" {
+			return false
+		}
+	}
+	return true
 }
 
 // ValidateMountMode will make sure the mount mode is valid.
 // returns if it's a valid mount mode and if it's read-write or not.
 func ValidateMountMode(mode string) (bool, bool) {
-	return roModes[mode] || rwModes[mode], rwModes[mode]
+	return validMode(mode), rwMode(mode)
 }
 
 // ReadOnly tells you if a mode string is a valid read-only mode or not.
 func ReadWrite(mode string) bool {
-	return rwModes[mode]
+	return rwMode(mode)
+}
+
+func rchown(fpath string, uid, gid int) error {
+	exclude_path := []string{"/", "/usr", "/etc"}
+	for _, p := range exclude_path {
+		if fpath == p {
+			return fmt.Errorf("Chowning of %s is not allowed", fpath)
+		}
+	}
+	callback := func(p string, info os.FileInfo, err error) error {
+		return os.Chown(p, uid, gid)
+	}
+
+	return filepath.Walk(fpath, callback)
+}
+
+func FixUidGid(userid, src, modes string) error {
+	chown_user := false
+	if userid == "" {
+		return nil
+	}
+
+	arr := strings.Split(modes, ",")
+	for _, mode := range arr {
+		if mode == "u" {
+			chown_user = true
+			break
+		}
+	}
+	if !chown_user {
+		return nil
+	}
+	// Set up defaults.
+	defaultExecUser := user.ExecUser{
+		Uid:  syscall.Getuid(),
+		Gid:  syscall.Getgid(),
+		Home: "/",
+	}
+	passwdPath, err := user.GetPasswdPath()
```

Comment:
This is the exact code copied from the handling of -u when processes are execed in libcontainer.  It reads the /etc/passwd from inside the container to determine the UID/GID of the container.  I believe it will get the correct UID/GID based on the container contents.